spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/PartTreeDatastoreQuery.java

```diff
@@ -108,36 +103,51 @@ else if (this.tree.isDistinct()) {
 	public Object execute(Object[] parameters) {
 		Supplier<StructuredQuery.Builder<?>> queryBuilderSupplier = StructuredQuery::newKeyQueryBuilder;
 		Function<T, ?> mapper = Function.identity();
-		Collector<?, ?, ?> collector;
+		Collector<?, ?, ?> collector = Collectors.toList();
+		Class returnedType = this.queryMethod.getReturnedObjectType();
 
-		if (this.tree.isCountProjection()) {
+		boolean returnedTypeisNumber = Number.class.isAssignableFrom(returnedType)
+				|| returnedType == int.class || returnedType == long.class;
+
+		boolean isCountingQuery = this.tree.isCountProjection()
+				|| (this.tree.isDelete() && returnedTypeisNumber);
+
+		if (isCountingQuery) {
 			collector = Collectors.reducing(0, e -> 1, Integer::sum);
 		}
 		else if (this.tree.isExistsProjection()) {
-			collector = Collectors.collectingAndThen(Collectors.counting(), count -> count > 0);
+			collector = Collectors.collectingAndThen(Collectors.counting(),
+					count -> count > 0);
 		}
-		else {
+		else if (!returnedTypeisNumber) {
 			queryBuilderSupplier = StructuredQuery::newEntityQueryBuilder;
 			mapper = this::processRawObjectForProjection;
-			collector = Collectors.toList();
 		}
 
 		StructuredQuery.Builder<?> structredQueryBuilder = queryBuilderSupplier.get();
 		structredQueryBuilder.setKind(this.datastorePersistentEntity.kindName());
 		applyQueryBody(parameters, structredQueryBuilder);
-		Iterable results = this.datastoreTemplate
+		Iterable rawResults = this.datastoreTemplate
 				.queryKeysOrEntities(structredQueryBuilder.build(),
 				this.entityType);
 
-		Object returned = results == null ? null
-				: StreamSupport.stream(results.spliterator(), false).map(mapper)
+		Object result = rawResults == null ? null
+				: StreamSupport.stream(rawResults.spliterator(), false).map(mapper)
 						.collect(collector);
 
-		return this.tree.isExistsProjection() || this.tree.isCountProjection()
-				|| results == null
-						? returned
+		if (this.tree.isDelete()) {
+			if (returnedTypeisNumber) {
+				this.datastoreTemplate.deleteAllById(rawResults, this.entityType);
+			}
+			else {
+				this.datastoreTemplate.deleteAll(rawResults);
+			}
+		}
+
+		return this.tree.isExistsProjection() || isCountingQuery || result == null
```

Comment:
I think there is way too much conditional logic in this method. There are like a dozen different if statements here. Can this method be broken up into smaller ones or reorganized to make it more readable and easier to maintain going forward?